### ISSUE TITLE
False negative E0308 for function with return  type `!`

### ISSUE THREAD

**Issue by neonaot** [intellij-rust/intellij-rust#8117]
## Environment

* **IntelliJ Rust plugin version:** 0.4.161.4249-213-nightly
* **Rust toolchain version:** 1.56.0 (09c42c458 2021-10-18) x86_64-apple-darwin
* **IDE name and version:** CLion 2021.3 Beta (CL-213.5744.5)
* **Operating system:** macOS 11.4
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
E0308 is not detected, despite that the loop is not endless:
<img width="699" alt="изображение" src="https://user-images.githubusercontent.com/49211026/142166518-8da51ab9-0216-49cc-916e-1f8e9fd02014.png">

## Steps to reproduce
```rust 
fn main() {
    foo(7);
}

fn foo(x: i32) -> ! {
    loop {
        if x == 7 {
            break
        }
    }
}

```
